File: packages/typescript/src/transport.ts
import type { YaiEnvelope } from './envelope';
import type { YaiOperationId } from './operations';
import type { YaiCallContext } from './call-context';

export interface YaiTransport {
  invoke<TData = unknown>(
    operationId: YaiOperationId,
    request?: unknown,
    callContext?: YaiCallContext
  ): Promise<YaiEnvelope<TData>>;
}

export class YaiTransportNotConfiguredError extends Error {
  constructor(operationId: string) {
    super(`YAI transport is not configured for operation ${operationId}`);
    this.name = 'YaiTransportNotConfiguredError';
  }
}

export function unavailableEnvelope<TData = unknown>(operationId: YaiOperationId, message: string): YaiEnvelope<TData> {
  return {
    operation_id: operationId,
    status: 'unavailable',
    execution_claim: false,
    implementation_status: 'transport-unconfigured',
    message
  };
}

File: packages/rust/src/transport.rs
use serde::de::DeserializeOwned;

use crate::call_context::YaiCallContext;
use crate::envelope::YaiEnvelope;
use crate::status::{YaiError, YaiResult};

pub trait YaiTransport {
    fn invoke<TData>(
        &self,
        operation_id: &'static str,
        request: Option<serde_json::Value>,
    ) -> YaiResult<YaiEnvelope<TData>>
    where
        TData: DeserializeOwned;

    fn invoke_with_context<TData>(
        &self,
        operation_id: &'static str,
        request: Option<serde_json::Value>,
        _context: YaiCallContext,
    ) -> YaiResult<YaiEnvelope<TData>>
    where
        TData: DeserializeOwned,
    {
        self.invoke(operation_id, request)
    }
}

#[derive(Debug, Default, Clone)]
pub struct NotConfiguredTransport;

impl YaiTransport for NotConfiguredTransport {
    fn invoke<TData>(
        &self,
        operation_id: &'static str,
        _request: Option<serde_json::Value>,
    ) -> YaiResult<YaiEnvelope<TData>>
    where
        TData: DeserializeOwned,
    {
        Err(YaiError::TransportNotConfigured { operation_id })
    }
}

#[cfg(test)]
mod tests {
    use super::{NotConfiguredTransport, YaiTransport};
    use crate::status::YaiError;

    #[test]
    fn not_configured_transport_returns_not_configured() {
        let transport = NotConfiguredTransport;
        let result = transport.invoke::<serde_json::Value>("system.status", None);
        assert!(matches!(
            result,
            Err(YaiError::TransportNotConfigured {
                operation_id: "system.status"
            })
        ));
    }
}
